design/authenticators/storing_non_sensitive_info.md

```diff
@@ -0,0 +1,331 @@
+# Store Authenticators non-sensitive information
+
+- [Store Authenticators non-sensitive information](#store-authenticators-non-sensitive-information)
+  * [Introduction](#introduction)
+  * [Solution](#solution)
+    + [Authenticator modification](#authenticator-modification)
+    + [Backwards compatibility](#backwards-compatibility)
+  * [Test plan](#test-plan)
+  * [Performance](#performance)
+    + [Configuration](#configuration)
+    + [Authentication](#authentication)
+  * [Security](#security)
+  * [Documentation](#documentation)
+  * [Version Update](#version-update)
+  * [Delivery Plan](#delivery-plan)
+  
+## Introduction
+
+In several authenticators we need to store some information for their
+authentication process. For example, in `authn-oidc` we have 2 variables
+that store this kind of data:
+  - provider-uri
+  - id-token-user-property
+  
+These 2 variables are defined in the authenticator policy and are loaded with
+values before users can authenticate with the authenticator. However, this data
+is not sensitive and was stored in variables just because we didn't think of a 
+better solution. As the data is not sensitive, it makes sense to store them in
```

Comment:
@orenbm why does it make sense to store them in annotations rather than variables? The variables can store any data, not just sensitive data. It provides a unified location where all the authenticator configuration is located - both sensitive and non-sensitive information would be stored in the same place. In your proposal, the user would have to understand which configuration should be in annotations and which in variables and use different APIs for each. In that aspect, I think the UX is less optimal.

What if we simplified the configuration of authenticators through new commands for the CLI?
For example, to define a new authenticator, call:
```bash
$ conjur config authenticator add authn-oidc/my-idp --provider-uri my-uri --id-token-user-property prop
```

Then to see the authenticator, a user could run:
```bash
$ conjur config authenticator show authn-oidc/my-idp
```
Update the authenticator with:
```bash
$ conjur config authenticator update authn-oidc/my-idp --provider-uri my-new-uri
```

And delete an authenticator with:
```bash
$ conjur config authenticator delete authn-oidc/mu-idp
```

Each command would perform the required policy manipulations behind the scenes. So one can either work with the policies directly as done today, or do it in simple commands through the CLI.

I think there are ways that we can improve the UX even further, without breaking the API and splitting the data into two different storage locations (annotations and variables).

What do you think?

Reply:
1. The user doesn't need to understand different APIs. He will load a policy with the non-sensitive data and will load sensitive data into variables. It also makes sense to store sensitive data in variables and non-sensitive data in another data-object, as variables hold secrets by definition.
2. As it's a nice feature to define authenticators via the CLI, it has nothing to do with the suggested approach. If we implement the suggested feature, then the implementation of the CLI command will be slightly different but it won't change anything from the user's perspective. We can do it in the future but I wouldn't treat these 2 features as related or dependant on each other. I also think that the implementation time, and testing time, of your suggestion is A LOT longer than the one suggested in this doc.

Reply:
Having all of the CLI commands would be beneficial. However I still think that non-sensitive authenticator stored as annotations makes configuring authenticators easier, less room for errors and easier to troubleshoot (On the support side of things).

For OIDC we turn 3 steps into 1 step.
For Azure we turn 2 steps into 1 step.

Reply:
I think we need to find the root cause of a problem, to solve it properly and make our product better. And not be tempted to make shortcuts that create technical debt. If we want to put information as annotation, because currently, it's easier to view them, that is a concern to me. If we want to put information in annotation because we believe that this is where this information absolutely should to be in, that's a different story. But I feel that currently, the answer is the former.

I am concerned that config information of the same instance (an authenticator) would be scattered between variables and annotations. While this saves 1-2 steps of commands, it requires the end user to hold a deeper understanding of the product. The user would need to know which parameter goes where and the user would have to use two different approaches (`update variable` and `load policy`) to perform the same logical action (provide authenticator config data). The user currently loads a policy anyway, however the policy load is logically defining the authenticator, then updating the variables is how you logically set the authenticator configuration. With your offer, the authenticator definition the two logical steps are mixed into 1.5 of the process is done with policy load (definition + half of the authenticator config), then the 0.5 remaining is completed with a variable update of the sensitive config (if exists).

Reply:
@AndrewCopeland in your comment:

> For OIDC we turn 3 steps into 1 step.
> For Azure we turn 2 steps into 1 step.

You are are reducing the number of API calls, but you are not reducing the number of steps to one, because with the suggestion above, you are adding an `if` step for the user. The user needs to think and understand, where does each parameter go? Is it sensitive or non sensitive? I think that adds complexity and requires more knowledge and familiarity with the product to get right. Same with the retrieval. Instead of getting all the config values in the same way, the user would have to show half of the configuration values with one command and the rest of the configuration values with another.

Lastly, there's also the permissions consideration. Setting configuration in annotations, requires different permissions than updating variables. You would no longer be able to give a user limited permission to only modify an existing authenticator. Every user would have to have create/delete/modify permissions always.

Reply:
>  While this saves 1-2 steps of commands, it requires the end user to hold a deeper understanding of the product. The user would need to know which parameter goes where and the user would have to use two different approaches (update variable and load policy) to perform the same logical action (provide authenticator config data)

> with the suggestion above, you are adding an if step for the user. The user needs to think and understand, where does each parameter go? Is it sensitive or non sensitive? I think that adds complexity and requires more knowledge and familiarity with the product to get right.

Same as above. this complexity is not in the customer's side.

> Same with the retrieval. Instead of getting all the config values in the same way, the user would have to show half of the configuration values with one command and the rest of the configuration values with another.

That's incorrect. The customer will never retrieve these secrets. They are defined for the use of the authenticator, which retrieves them during the authentication process. So again, no added complexity for the customer.
That is not correct. _We_ will defined the authenticator and we will decide wether the data is sensitive or not. If a community member will create the authenticator then they will need to decide that. The end-user doesn't need to understand that as the docs will state which data should go in annotations and which data should go in variables. I actually think that it makes sense to tell the user to store metadata in annotations and secrets in variables. 

> Lastly, there's also the permissions consideration. Setting configuration in annotations, requires different permissions than updating variables. You would no longer be able to give a user limited permission to only modify an existing authenticator. Every user would have to have create/delete/modify permissions always.

What do you mean here? Currently for a user to update a configuration he will need `update` permissions on the variable. After this change he will need `update` permissions on the webservice. Do you see a scenario in which a user will get permissions on the variables of the authenticator and not on the webservice? it makes no sense.

Reply:
>The end-user doesn't need to understand that as the docs will state which data should go in annotations and which data should go in variables

Going to the docs is not a great user experience. But running a CLI that completes your required and optional parameters is. Because it relieves you from the need to rely on the documentation and prevents you from making any unnecessary human errors, such as parameter typos.

Reply:
> Going to the docs is not a great user experience. But running a CLI that completes your required and optional parameters is

To configure/enable any authenticator I expect the user to go to the docs, I do not expect a customer to understand how to configure authenticators without going to the docs. I understand your point of view for having the conjur-cli enable the authenticator but this brings up another concern which is policy as code. If we enable authenticators with the conjur-cli we do not have a record of which authenticators are configured within the policy.

> You are are reducing the number of API calls, but you are not reducing the number of steps to one, because with the suggestion above, you are adding an if step for the user. The user needs to think and understand, where does each parameter go?

I do not see any `if` step here because the docs **should** walk the user through which annotations need to be set and then which variables need to set.

> If we want to put information as annotation, because currently, it's easier to view them, that is a concern to me. If we want to put information in annotation because we believe that this is where this information absolutely should to be in, that's a different story. 

I think it's both, it's easier to view and update also it is where non-sensitive information should reside.